engine/crates/engine-v2/src/execution/planner/partition.rs

```diff
@@ -0,0 +1,775 @@
+use std::borrow::Cow;
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub(super) enum Partition<Id> {
+    One(Id),
+    Many(Vec<Id>),
+}
+
+impl<Id> From<Id> for Partition<Id> {
+    fn from(id: Id) -> Self {
+        Partition::One(id)
+    }
+}
+
+impl<Id: Copy> From<Vec<Id>> for Partition<Id> {
+    fn from(ids: Vec<Id>) -> Self {
+        match ids[..] {
+            [] => unreachable!(),
+            [id] => Partition::One(id),
+            _ => Partition::Many(ids),
+        }
+    }
+}
+
+/// For selection sets with type conditions, we must know the concrete object id being
+/// resolved to apply correctly the field collection as defined in the specification:
+/// https://spec.graphql.org/October2021/#CollectFields()
+///
+/// However, we're building the expect shape of the subgraph response in advance to drive the
+/// de-serialization, so we don't have it! So we need to pre-compute all possible shapes and
+/// use the right one later.
+///
+/// This function will create a partition set of the output object ids that guarantees
+/// the following:
+/// - Each partition is disjoint.
+/// - All object ids in a partition will have the same fields.
+///
+/// To do so it's given:
+/// - output_ids: the list of possible object ids from that the parent field may return
+/// - type_conditions: All possible type conditions, each treated as a list of possible object ids.
```

Comment:
Again, maybe could be framed in terms of the current selection set?

"Each type condition in the selection set" rather than "all possible"?  Maybe also worth clarifying that "possible object ids" means the object ids that would match the given type condition?  (Kind of implied by what you've got, but not sure how obvious it was).